#33 feat : 친구 - 친구 조회 및 삭제

yunah7 opened the pull request:
## 요약
- 컨벤션 처리
- 예외 추가 및 static import로 변경
- sql count 메소드로 connector 수 반환
- 친구 리스트 조회 및 최근 교류 친구 조회 기능 추가

## 변경 내용
**1. Controller**
```
	@PostMapping("/count")
	public int countConnection() {
		return connectionService.countConnection();
	}
       
       @PostMapping("/delete/{connectorUserId}")
	public void deleteConnection(@PathVariable String connectorUserId) {
		connectionService.deleteConnection(connectorUserId);
	}

	@PostMapping("/find")
	public List<String> findTenConnection() {
		return connectionService.findTenConnection();
	}

	@PostMapping("/find-all")
	public List<String> findAllConnection() {
		return connectionService.findAllConnection();
	}

	@PostMapping("/find-recent")
	public List<String> findRecentConnection() {
		return connectionService.findRecentConnection();
	}
 
```
- 친구 수 조회, 친구 삭제, 조회(기본적으론 10명만 조회, 더보기 클릭 시 전체 조회), 가장 최근에 교류한 친구 3명 조회
- 조회 시 String List로 userId들을 전달하기에, 삭제 시에도 userId를 받아오도록 함

**2. ConnectionService**
```
public int countConnection() {
		int loginId = loginService.getLoginSession();
		return connectionMapper.countConnection(loginId);
	}

	public void deleteConnection(String connectorUserId) {
		int loginId = loginService.getLoginSession();
		int connectorId = userService.getIdByUserId(connectorUserId);
		if (loginId == connectorId) {
			throw new AuthorizationException(SELF_DISCONNECTION);
		}
		Connection connection = connectionMapper.getConnection(loginId, connectorId);
		connectionMapper.modifyActivate(connection.getId(), false);
	}

	public List<String> findTenConnection() {
		int loginId = loginService.getLoginSession();
		if (connectionMapper.countConnection(loginId) == 0) {
			throw new AuthorizationException(NO_CONNECT_ANYONE);
		}
		return connectionMapper.findTenConnection(loginId);
	}

	public List<String> findAllConnection() {
		int loginId = loginService.getLoginSession();
		return connectionMapper.findAllConnection(loginId);
	}

	public List<String> findRecentConnection() {
		int loginId = loginService.getLoginSession();
		if (connectionMapper.countConnection(loginId) == 0) {
			throw new AuthorizationException(NO_CONNECT_ANYONE);
		}
		return connectionMapper.findRecentConnection(loginId);
	}
```

- 재활용성을 위해 유저의 id를 가져오고, id를 통해 connection을 조회하도록 함
- 기본적으로 세션에 저장된 USER 테이블 row의 id를 얻어와 각 조건에 맞는 예외 처리 후 mapper을 통한 DB연결

**3. ConnectionMapper.xml**
```
    <select id="countConnection" resultType="int">
        SELECT count(*)
        FROM CONNECTOR
        WHERE (connector_id = #{userId} OR connectee_id = #{userId})
          AND (CONNECTOR.activate = 1);
    </select>

    <select id="findTenConnection" resultType="String">
        SELECT USER.user_id
        FROM CONNECTOR
                 JOIN USER ON CONNECTOR.connector_id = USER.id OR CONNECTOR.connectee_id = USER.id
        WHERE ((CONNECTOR.connector_id = #{userId} OR CONNECTOR.connectee_id = #{userId}) AND
               USER.id
            != #{userId}
          AND CONNECTOR.activate = 1 ) LIMIT 10;

    </select>

    <select id="findAllConnection" resultType="String">
        SELECT USER.user_id
        FROM CONNECTOR
                 JOIN USER ON CONNECTOR.connector_id = USER.id OR CONNECTOR.connectee_id = USER.id
        WHERE (CONNECTOR.connector_id = #{userId} OR CONNECTOR.connectee_id = #{userId})
          AND USER.id != #{userId}
          AND
            CONNECTOR.activate = 1;
    </select>

    <select id="findRecentConnection" resultType="String">
        SELECT USER.user_id
        FROM CONNECTOR
                 JOIN USER ON CONNECTOR.connector_id = USER.id OR CONNECTOR.connectee_id = USER.id
        WHERE (CONNECTOR.connector_id = #{userId} OR CONNECTOR.connectee_id = #{userId})
          AND USER.id != #{userId}
          AND
            CONNECTOR.activate = 1
        ORDER BY CONNECTOR.last_visited_at DESC
            LIMIT 3;
    </select>
```

- 매개변수로 받은 userId(USER TABLE의 index)값이 CONNECTOR table의 connector_id나 connectee_id 값에 위치하고, 
  CONNECTOR table과 USER table을 JOIN한 row 중 USER table의 id가 userId가 아니며, 
  CONNECTOR table의 activate가 true인 값 
  = 활성화된 친구 리스트 

## Issue number/URL
#33

yunah7 committed:
#32 feat : 친구 추가 및 초대 코드 생성

RandomStringUtils 내 randomAlphabetic을 통한 랜덤 문자열로 초대코드 생성
PathVariable 어노테이션을 통해 초대코드로 친구 추가

yunah7 committed:
#33 feat : 친구 - 친구 조회 및 삭제

컨벤션 처리
예외 추가 및 static import로 변경
sql count 메소드로 connector 수 반환
친구 리스트 조회 및 최근 교류 친구 조회 기능 추가

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -24,4 +26,29 @@ public void createConnection(@PathVariable String connectorCode) {
+		return connectionService.countConnection();
+	}
+
+	@PostMapping("/delete/{connectorUserId}")
+	public void deleteConnection(@PathVariable String connectorUserId) {
+		connectionService.deleteConnection(connectorUserId);
+	}
```
이런건 DeleteMapping을 사용해서 uri에 delete를 생략할 수도 있겠네요!

sunwootest commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -24,4 +26,29 @@ public void createConnection(@PathVariable String connectorCode) {
+		connectionService.deleteConnection(connectorUserId);
+	}
+
+	@PostMapping("/find")
+	public List<String> findTenConnection() {
+		return connectionService.findTenConnection();
+	}
+
+	@PostMapping("/find-all")
+	public List<String> findAllConnection() {
+		return connectionService.findAllConnection();
+	}
+
+	@PostMapping("/find-recent")
+	public List<String> findRecentConnection() {
+		return connectionService.findRecentConnection();
+	}
```
이것들도 getMapping에 가까워보이네요 :) 그러면 find를 생략할 수도 있겠고요 !

uri010 commented on src/main/java/study/sharedcalendar/constant/ErrorCode.java:
```diff
@@ -14,13 +14,14 @@ public enum ErrorCode {
 	EXCEEDED_PASSWORD_VALIDITY_PERIOD(HttpStatus.FORBIDDEN, "비밀번호 변경 주기가 지났습니다. 비밀번호를 재설정해주세요."),
-	SELF_INVITATION(HttpStatus.FORBIDDEN, "자기 자신에게는 초대를 받을 수 없습니다."),
+	SELF_CONNECTION(HttpStatus.FORBIDDEN, "자기 자신에게는 초대를 받을 수 없습니다."),
+	SELF_DISCONNECTION(HttpStatus.FORBIDDEN, "자기 자신에게는 삭제를 할 수 없습니다."),
```
메시지를 확인하지 않으면 유추하기 조금 힘든 것 같습니다. 좀 더 적절한 표현 부탁드립니다!

uri010 commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -24,4 +26,29 @@ public void createConnection(@PathVariable String connectorCode) {
+	@PostMapping("/find-all")
+	public List<String> findAllConnection() {
+		return connectionService.findAllConnection();
+	}
```
Connector을 조회할 때 처음 10명을 보여주고, 더보기 클릭시 10명을 더 보여주는 식으로 요구사항을 정했던 걸로 기억합니다. 전체 Connector 조회가 아닌 10명을 조회하고 그 다음 10명을 조회하는 식으로 수정 부탁드립니다.

yunah7 replied:
redis를 통해 offset을 저장하는 방식으로 수정했었는데, redis 서버 오류 시, 해당 서비스에도 문제가 나타날 수 있어, 클라이언트에서 전체 유저의 정보를 전달하는 방식으로 수정해두었습니다 !

uri010 commented on src/main/resources/mybatis-mapper/ConnectionMapper.xml:
```diff
@@ -6,25 +6,66 @@
     </insert>
+
+    <select id="countConnection" resultType="int">
+        SELECT count(*)
```
테이블명과 sql 문법은 대문자로 표현하기로 정한 거 같은데 count의 경우도 대문자로 수정하는 게 좋을 것 같아요.

uri010 commented on src/main/resources/mybatis-mapper/ConnectionMapper.xml:
```diff
@@ -6,25 +6,66 @@
+        SELECT USER.user_id
+        FROM CONNECTOR
+                 JOIN USER ON CONNECTOR.connector_id = USER.id OR CONNECTOR.connectee_id = USER.id
+        WHERE ((CONNECTOR.connector_id = #{userId} OR CONNECTOR.connectee_id = #{userId}) AND
+               USER.id
+            != #{userId}
```
그리고 sql문이 긴 경우는 줄바꿈을 적절히 사용해주세요! 
ithub에선 조금 다르게 보여지는 거 같은데 AND / USER.id / !=#{userId}마다 줄바꿈이 되어 있네요. IDE에서 확인 부탁드려요!

yunah7 replied:
문장이 길다보니 적용한 컨벤션에서 이런 식으로 자동 정렬을 해주는 것으로 보입니다..! 해당 부분은 수동으로 수정하겠습니다 :)

yunah7 committed:
Merge branch 'feature/32' into feature/33

# Conflicts:
#	src/main/java/study/sharedcalendar/constant/ErrorCode.java
#	src/main/java/study/sharedcalendar/constant/ErrorResponse.java
#	src/main/java/study/sharedcalendar/constant/UserConstant.java
#	src/main/java/study/sharedcalendar/controller/ConnectionController.java
#	src/main/java/study/sharedcalendar/dto/Connection.java
#	src/main/java/study/sharedcalendar/dto/LoginReq.java
#	src/main/java/study/sharedcalendar/dto/SignUpReq.java
#	src/main/java/study/sharedcalendar/dto/User.java
#	src/main/java/study/sharedcalendar/exception/AuthorizationException.java
#	src/main/java/study/sharedcalendar/exception/DuplicateException.java
#	src/main/java/study/sharedcalendar/exception/GlobalExceptionHandler.java
#	src/main/java/study/sharedcalendar/exception/NoMatchedConnectionException.java
#	src/main/java/study/sharedcalendar/exception/NoMatchedUserException.java
#	src/main/java/study/sharedcalendar/mapper/ConnectionMapper.java
#	src/main/java/study/sharedcalendar/mapper/LoginMapper.java
#	src/main/java/study/sharedcalendar/mapper/UserMapper.java
#	src/main/java/study/sharedcalendar/service/BcryptEncryptionService.java
#	src/main/java/study/sharedcalendar/service/ConnectionService.java
#	src/main/java/study/sharedcalendar/service/EncryptionService.java
#	src/main/java/study/sharedcalendar/service/LoginService.java
#	src/main/java/study/sharedcalendar/service/UserService.java
#	src/main/resources/ddl.sql
#	src/main/resources/mybatis-mapper/ConnectionMapper.xml
#	src/main/resources/mybatis-mapper/LoginMapper.xml
#	src/main/resources/mybatis-mapper/UserMapper.xml

yunah7 committed:
#33 fix : 친구 - no-offest 페이징 처리를 통한 조회

Connection에 관련된 상수를 ConnectionConstant를 통해 처리
컨트롤러 어노테이션 수정
no-offset 페이징 처리 - redis를 통해 마지막 조회 id 저장

yunah7 committed:
#33 fix : offset 획득 방식 수정

redis를 통해 마지막 조회유저를 얻어왔던 방식에서, 클라이언트에서 조회한 전체 유저의 정보를 넘겨주는 방식으로 수정

yunah7 committed:
#33 fix : 불필요한 데이터 삭제

ddl.sql에 넣어두었던 테스트 데이터 삭제

yunah7 commented:
## 변경 사항

**1. offset 획득 방식 수정**
```java
	public List<String> findConnection(List<String> connection) {
		int loginId = loginService.getLoginSession();
		if (connectionMapper.countConnection(loginId) == 0) {
			throw new AuthorizationException(NO_CONNECT_ANYONE);
		}

		int offset;
		if (connection == null) {
			offset = 0;
		} else {
			String lastConnection = connection.get(connection.size() - 1);
			offset = connectionMapper.getConnection(loginId,
				userService.getIdByUserId(lastConnection)).getId();
		}

		List<String> connectionList = connectionMapper.findConnection(loginId, offset);

		if (connectionList.isEmpty()) {
			throw new AuthorizationException(NO_VIEW_ANYONE);
		}
		return connectionList;
	}
```
- redis를 통해 마지막 조회 유저를 저장했던 방식에서, 클라이언트에서 현재까지 조회한 모든 정보의 유저를 넘겨주도록 수정
  - null값을 받아올 수 있도록 Controller의 RequestBody 특성에 requrired=false 조건 추가
  - null값을 받아왔을 때에는 offset을 0으로, 조회한 유저의 리스트를 받아왔을 때에는 가장 마지막으로 넘어온 유저의 id를 offset으로 설정

uri010 commented on src/main/java/study/sharedcalendar/constant/ErrorCode.java:
```diff
@@ -14,17 +14,21 @@ public enum ErrorCode {
 	EXCEEDED_LOGIN_ATTEMPTS(HttpStatus.FORBIDDEN, "로그인 횟수가 초과했습니다. 비밀번호를 재설정해주세요."),
 	EXCEEDED_PASSWORD_VALIDITY_PERIOD(HttpStatus.FORBIDDEN, "비밀번호 변경 주기가 지났습니다. 비밀번호를 재설정해주세요."),
-	SELF_INVITATION(HttpStatus.FORBIDDEN, "자기 자신에게는 초대를 받을 수 없습니다."),
+	SELF_CONNECTION_REQUEST(HttpStatus.FORBIDDEN, "자기 자신에게는 초대를 받을 수 없습니다."),
+	SELF_DISCONNECTION_REQUEST(HttpStatus.FORBIDDEN, "자기 자신에게는 삭제를 할 수 없습니다."),
```
초대와 삭제를 못한다는 의미로 not possible이나 Cannot와 같은 메시지를 붙이는 건 어떤가요?

uri010 commented on src/main/java/study/sharedcalendar/controller/ConnectionController.java:
```diff
@@ -28,4 +33,24 @@ public void changeInviteUrlCode() {
+		connectionService.deleteConnection(connectorUserId);
+	}
+
+	@GetMapping()
```
GetMapping 다음에 오는 괄호는 없애주셔도 됩니다!

uri010 reviewed (approved):
approve 드립니당
